catalog/app/testing/util.js

```diff
@@ -1,17 +1,143 @@
+ *
+ * @param {string[]|function} drop
+ *
+ * @returns {function} Attribute mapper.
+ */
+const getAttrs = (mapper, drop) => {
+  if (typeof mapper === 'function') return mapper;
+  return flow(
+    drop instanceof Array && !mapper ? omit(drop) : id,
+    mapper instanceof Array ? pick(mapper) : mapper || id,
+  );
+};
 
-export const findMockComponent = (html, name) =>
-  html.find(`[__name="${name}"]`);
+const defaultRenderChildren = ({ children }) => [children];
 
-export const flushPromises = () => new Promise((resolve) => setImmediate(resolve));
+const renderChildren = (props) =>
+  Object.entries(props).map(([key, value]) =>
+    createElement(getPropName(key), {}, value));
+
+/**
+ * Get children from props.
+ *
+ * @param {Object} props
+ *
+ * @param {string[]|function} mapper
+ *
+ * @returns {react.Element[]} An array of react elements to use as children.
+ */
+const getChildren = (mapper) =>
+  mapper instanceof Array
+    ? flow(pick(mapper), pickBy(Boolean), renderChildren)
+    : mapper || defaultRenderChildren;
+
+/**
+ * Create a mock stateless functional component that renders as a custom html element.
+ *
+ * @param {string} name
+ *
+ * @param {Object} options
+ *
+ * @param {string[]|function} options.attrs
+ *   If array of strings, pick these props and render them as resulting element's attributes.
+ *   If function, call it with props and use returned object as resulting element's attributes.
+ *   Renders all the props by default.
+ *
+ * @param {string[]|function} options.children
+ *   If array of strings, pick these props and render them as nested elements, like this:
+ *     <prop-${key}>{value}</prop-${key}>
+ *   If function, call it with props and use returned array as resulting element's children.
+ *   Renders children as they are by default.
+ *
+ * @returns {react.Component}
+ */
+export const mockComponent = (name, {
+  attrs,
+  children,
+} = {}) => {
+  const attrMapper = getAttrs(attrs, children);
+  const childrenMapper = getChildren(children);
+  const component = getComponentName(name);
+  return (props) => createElement(component, attrMapper(props), ...childrenMapper(props));
+};
+
+/**
+ * Create a mock class component that renders as a custom html element.
+ * The same as `mockComponent`, but additionally accepts a method map that is
+ * merged into prototype.
+ *
+ * @param {string} name
+ *
+ * @param {Object} options
+ *
+ * @param {Object} options.methods A map of methods to merge into prototype.
+ *
+ * @returns {react.Component}
+ */
+export const mockComponentClass = (name, { methods = {}, ...opts }) => {
+  const render = mockComponent(name, opts);
+
+  class Mock extends Component {
+    render() { return render(this.props); }
+  }
+
+  Object.assign(Mock.prototype, methods);
+
+  return Mock;
+};
+
+/**
+ * Create a selector (suitable for enzyme or cheerio wrappers) that selects
+ * a mock component created by `mockComponent[Class]` or its nested element / prop.
+ *
```

Comment:
do we have to have an empty line between `@param` calls? it's not a big deal but we should be consistent across the project.

Reply:
I usually put an empty line if params are not related and/or they have multiline description. If params are related (example: object props) and have short descriptions, I usually don't put an empty line.